rfd/0212-jsonpath-interpolation.md

```diff
@@ -0,0 +1,635 @@
+basics of [JSONPath syntax](https://support.smartbear.com/alertsite/docs/monitors/api/endpoint/jsonpath.html).
+
+When going through the JSONPath examples, you may find it useful to run the
+queries in a [sandbox](https://serdejsonpath.live/).
+
+#### `jsonpath` Expression Function
+
+The `jsonpath` function will be added as a trait expression function, but will
+only be supported for `login_rule` expressions (`traits_map` or `traits_expression`).
+
+It can be used to interpolate a string or strings from arbitrary JSON claims.
+For example, see the JSON object and jsonpath examples below:
+
+```json
+{
+  "a": ["1", "2", "3"],
+  "b": {
+    "c": "d"
+  }
+}
+```
+
+* `jsonpath("$.a")`   -> `["1", "2", "3"]`
+* `jsonpath("$.b.*")` -> `["d"]`
+* `jsonpath("$.*.*")` -> `["1", "2", "3", "d"]`
+
+#### JSONPath Libraries
+
+While the JSONPath query language has an [official RFC](https://www.rfc-editor.org/rfc/rfc9535.html)
+as of 2024, the inception of the query language was in an [article](https://goessner.net/articles/JsonPath/)
+written in 2007. The original article left many questions unanswered, causing
+the JSONPath language to morph in several directions as different JSONPath
+projects came up with their own answers.
+
+While the RFC has helped to realign these disparate projects, there is much
+more work to be done, as can be clearly seen by this [comparison project](https://cburgmer.github.io/json-path-comparison/).
+
+According to the comparison project, https://github.com/ohler55/ojg is
+currently the closest to the RFC specification out of the Go projects listed
+there.
+
+By using a project close to the specification, we can selectively rely on
+official JSONPath documentation, like those linked above, as well as the
+[online sandboxes](https://serdejsonpath.live/).
+
+Note: We must be careful when updating the upstream library, as some of the
+syntax may be subject to change as it adjusts to the RFC and some of the
+unresolved syntax disagreements in the community.
+
+### OIDC Claims
+
+During OIDC login, a user's OIDC claims serve two purposes:
+
+* Setting claims as Teleport user traits, optionally using login rules for custom claims to traits mapping.
+* Determining what Teleport roles to give the user, using the OIDC connector's `claims_to_roles` field to map user traits to roles.
+
+The `jsonpath` function will be supported in login rule trait mapping. As a
+result, any JSON OIDC claims mapped to traits in the claims to traits mapping
+will be available in the traits to roles mapping.
+
+Note: see [this section](#json-traits) for an explanation as to why we decided
+not to set arbitrary JSON values as user traits directly, and instead are
+requiring the use of login rules.
+
+#### Login Rules - Claims to Traits
+
+You can use a login rule to map JSON claims to users traits.
+
+In the example below, a JSON claim is mapped to individual user traits.
+
+```json
+{
+  // groups is a JSON object rather than a string array.
```

Comment:
nit: This is invalid JSON and displays strangely on GitHub with dark red background, white text. It also displays this way when merged I believe. Maybe just use ` ```js`?